File: src/stxm/io.py
import re
from pathlib import Path

import numpy as np


def _parse_points_count(text: str, axis_name: str) -> int | None:
    pattern = rf"{axis_name}\s*=\s*\{{[^}}]*Points\s*=\s*\(\s*(\d+)"
    match = re.search(pattern, text, re.DOTALL)
    return int(match.group(1)) if match else None


def _parse_points_array(text: str, axis_name: str) -> np.ndarray | None:
    pattern = rf"{axis_name}\s*=\s*\{{[^}}]*Points\s*=\s*\(\s*\d+\s*,\s*([\d\s.,\-eE+]+)\)"
    match = re.search(pattern, text, re.DOTALL)
    if not match:
        return None
    raw = match.group(1).replace(",", " ").split()
    return np.array([float(x) for x in raw])


def read_hdr(path: str | Path) -> dict:
    """
    Parse STXM .hdr file. Returns dict with axis sizes and optional axis arrays.

    Returns
    -------
    dict
        Keys: paxis_count, qaxis_count (int); paxis_name, qaxis_name (str);
        paxis_points, qaxis_points (ndarray, optional); raw (str).
    """
    path = Path(path)
    raw = path.read_text()
    paxis_count = _parse_points_count(raw, "PAxis")
    qaxis_count = _parse_points_count(raw, "QAxis")
    if paxis_count is None or qaxis_count is None:
        raise ValueError("Could not find PAxis or QAxis Points in header")
    out = {
        "paxis_count": paxis_count,
        "qaxis_count": qaxis_count,
        "raw": raw,
    }
    pname = re.search(r'PAxis\s*=\s*\{\s*Name\s*=\s*"([^"]*)"', raw)
    qname = re.search(r'QAxis\s*=\s*\{\s*Name\s*=\s*"([^"]*)"', raw)
    if pname:
        out["paxis_name"] = pname.group(1)
    if qname:
        out["qaxis_name"] = qname.group(1)
    parr = _parse_points_array(raw, "PAxis")
    qarr = _parse_points_array(raw, "QAxis")
    if parr is not None:
        out["paxis_points"] = parr
    if qarr is not None:
        out["qaxis_points"] = qarr
    return out


def read_xim(path: str | Path, shape: tuple[int, int] | None = None) -> np.ndarray:
    """
    Load STXM .xim ascii image (whitespace-separated values, one line per row).

    Parameters
    ----------
    path : str or Path
        Path to .xim file.
    shape : tuple of (n_rows, n_cols), optional
        If provided, ensure output has this shape (reshape by row-major if needed).
        If None, return 2D array as in file.

    Returns
    -------
    np.ndarray
        2D array of dtype float64. Shape (qaxis_count, paxis_count) when
        used with read_hdr dimensions.
    """
    path = Path(path)
    data = np.loadtxt(path, dtype=np.float64)
    if data.ndim == 1:
        if shape is None:
            raise ValueError("xim is 1D; provide shape=(n_rows, n_cols)")
        data = data.reshape(shape)
    elif shape is not None and data.size == np.prod(shape):
        if data.shape != shape:
            data = data.reshape(shape)
    return data


def load_stxm(
    hdr_path: str | Path,
    xim_path: str | Path | None = None,
) -> tuple[dict, np.ndarray]:
    """
    Load STXM scan: parse .hdr and load associated .xim ascii image.

    Parameters
    ----------
    hdr_path : str or Path
        Path to .hdr file.
    xim_path : str or Path, optional
        Path to .xim file. If None, inferred as same stem with '_a.xim'.

    Returns
    -------
    header : dict
        Result of read_hdr(hdr_path).
    image : np.ndarray
        2D array shape (qaxis_count, paxis_count), i.e. (n_rows, n_cols).
    """
    hdr_path = Path(hdr_path)
    meta = read_hdr(hdr_path)
    if xim_path is None:
        stem = hdr_path.stem
        parent = hdr_path.parent
        xim_path = parent / f"{stem}_a.xim"
        if not xim_path.exists():
            xim_path = parent / f"{stem}.xim"
    xim_path = Path(xim_path)
    if not xim_path.exists():
        raise FileNotFoundError(f"xim file not found: {xim_path}")
    shape = (meta["qaxis_count"], meta["paxis_count"])
    image = read_xim(xim_path, shape=shape)
    return meta, image


MIN_BYTES_PER_VALUE = 3

NEXAFS_LINE_SCAN_TYPE = 'Type = "NEXAFS Line Scan"'


def is_nexafs_line_scan_type(hdr_path: str | Path) -> bool:
    """
    Return True if the .hdr header contains Type = "NEXAFS Line Scan".

    Parameters
    ----------
    hdr_path : str or Path
        Path to .hdr file.

    Returns
    -------
    bool
    """
    path = Path(hdr_path)
    if not path.suffix.lower() == ".hdr" or not path.exists():
        return False
    return NEXAFS_LINE_SCAN_TYPE in path.read_text()


def is_valid_line_scan_fast(hdr_path: str | Path) -> bool:
    """
    Fast heuristic: True if .hdr parses with PAxis/QAxis and the .xim file exists
    and has at least (paxis_count * qaxis_count) * MIN_BYTES_PER_VALUE bytes.
    Use to filter file lists without loading full .xim data. May have false
    negatives (smaller .xim files that still load). Prefer is_valid_line_scan
    when a definitive result is needed.

    Parameters
    ----------
    hdr_path : str or Path
        Path to .hdr file.

    Returns
    -------
    bool
        True if header format matches and .xim size suggests full data.
    """
    try:
        hdr_path = Path(hdr_path)
        meta = read_hdr(hdr_path)
        n = meta["paxis_count"] * meta["qaxis_count"]
        stem = hdr_path.stem
        parent = hdr_path.parent
        xim_path = parent / f"{stem}_a.xim"
        if not xim_path.exists():
            xim_path = parent / f"{stem}.xim"
        if not xim_path.exists():
            return False
        return xim_path.stat().st_size >= n * MIN_BYTES_PER_VALUE
    except Exception:
        return False


def is_valid_line_scan(hdr_path: str | Path) -> bool:
    """
    Return True if the .hdr has an associated .xim that loads as a valid line scan
    (correct format and shape). Use to filter the file dropdown to only show loadable scans.

    Parameters
    ----------
    hdr_path : str or Path
        Path to .hdr file.

    Returns
    -------
    bool
        True if load_stxm(hdr_path) succeeds.
    """
    try:
        load_stxm(hdr_path)
        return True
    except Exception:
        return False


def is_nexafs_line_scan(hdr_path: str | Path) -> bool:
    """
    Return True if the .hdr is a NEXAFS Line Scan (Type = "NEXAFS Line Scan")
    and has an associated .xim that loads with the correct 2D line shape.

    Parameters
    ----------
    hdr_path : str or Path
        Path to .hdr file.

    Returns
    -------
    bool
        True if header type is NEXAFS Line Scan and load_stxm(hdr_path) succeeds.
    """
    return is_nexafs_line_scan_type(hdr_path) and is_valid_line_scan(hdr_path)


def _axis_is_energy(name: str) -> bool:
    lowered = name.lower()
    return "energy" in lowered or lowered.endswith("ev") or "photon" in lowered


def _is_strictly_monotonic(axis: np.ndarray) -> bool:
    diffs = np.diff(np.asarray(axis, dtype=np.float64).ravel())
    return bool(diffs.size) and bool(np.all(diffs > 0) or np.all(diffs < 0))


def orient_scan(meta: dict, image: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Return ``(energy, spatial, image)`` with shape ``(n_spatial, n_energy)``.

    Parameters
    ----------
    meta : dict
        Header dict from ``read_hdr`` with ``paxis_points`` and ``qaxis_points``.
    image : np.ndarray
        Raw 2D array from ``load_stxm``.

    Returns
    -------
    energy : np.ndarray
        Energy axis in eV, strictly monotonic.
    spatial : np.ndarray
        Spatial axis coordinates, length ``image.shape[0]`` after orientation.
    image : np.ndarray
        Oriented intensities ``(n_spatial, n_energy)``.

    Raises
    ------
    ValueError
        If axis sizes disagree with the array or energy is not monotonic.
    """
    paxis = np.asarray(meta["paxis_points"], dtype=np.float64)
    qaxis = np.asarray(meta["qaxis_points"], dtype=np.float64)
    arr = np.asarray(image, dtype=np.float64)
    p_name = str(meta.get("paxis_name", "PAxis"))
    q_name = str(meta.get("qaxis_name", "QAxis"))
    energy_on_p = _axis_is_energy(p_name)
    energy_on_q = _axis_is_energy(q_name)
    if arr.shape == (qaxis.size, paxis.size):
        spatial, energy = qaxis, paxis
    elif arr.shape == (paxis.size, qaxis.size):
        arr = arr.T
        spatial, energy = qaxis, paxis
    else:
        raise ValueError(
            f"image shape {arr.shape} incompatible with qaxis={qaxis.size} paxis={paxis.size}"
        )
    if energy_on_q and not energy_on_p:
        arr = arr.T
        spatial, energy = paxis, qaxis
    if not _is_strictly_monotonic(energy):
        raise ValueError("energy axis must be strictly monotonic")
    if arr.shape != (spatial.size, energy.size):
        raise ValueError("oriented image shape does not match axis lengths")
    return energy, spatial, arr


def spectral_matrix(
    image: np.ndarray,
    mask: np.ndarray | None = None,
    *,
    use_od: bool = True,
    eps: float = 1e-10,
) -> tuple[np.ndarray, tuple[int, int]]:
    """
    Flatten an oriented scan to ``(n_pixels, n_energy)``.

    Parameters
    ----------
    image : np.ndarray
        Oriented scan ``(n_spatial, n_energy)``.
    mask : np.ndarray, optional
        Boolean spatial mask; all rows used when ``None``.
    use_od : bool
        If True, transform to ``-ln(I)``; otherwise use intensities.
    eps : float
        Intensity floor when ``use_od`` is True.

    Returns
    -------
    X : np.ndarray
        Data matrix for demixing or regression.
    spatial_shape : tuple of int
        ``(n_spatial, 1)`` for reshaping abundances back to a line scan.

    Raises
    ------
    ValueError
        If the mask selects no rows.
    """
    arr = np.asarray(image, dtype=np.float64)
    if mask is None:
        block = arr
    else:
        mask = np.asarray(mask, dtype=bool)
        if mask.shape != (arr.shape[0],):
            raise ValueError("mask length must match n_spatial")
        block = arr[mask, :]
    if block.size == 0:
        raise ValueError("mask selects no spatial rows")
    if use_od:
        block = -np.log(np.maximum(block, eps))
    return block, (arr.shape[0], 1)


def list_nexafs_line_scans(experiment_path: str | Path) -> list[Path]:
    """
    List .hdr files in the experiment directory that are NEXAFS Line Scan type
    and have a loadable .xim with correct 2D shape. Sorted by name.

    Parameters
    ----------
    experiment_path : str or Path
        Path to experiment folder (contains .hdr and .xim files).

    Returns
    -------
    list of Path
        Paths to .hdr files that pass is_nexafs_line_scan.
    """
    directory = Path(experiment_path)
    if not directory.is_dir():
        return []
    return sorted(
        p for p in directory.glob("*.hdr")
        if is_nexafs_line_scan(p)
    )

File: src/lib/stxm/io.ts
import "server-only";

import fs from "node:fs";
import path from "node:path";

import type { ScanCategory } from "@/lib/stxm-types";

export const NEXAFS_LINE_SCAN_TYPE = 'Type = "NEXAFS Line Scan"';
export const MIN_BYTES_PER_VALUE = 3;

const HDR_TYPE_PATTERN = /Type\s*=\s*"([^"]*)"/;

export type HdrMeta = {
  paxis_count: number;
  qaxis_count: number;
  raw: string;
  paxis_name?: string;
  qaxis_name?: string;
  paxis_points?: number[];
  qaxis_points?: number[];
  energy_eV?: number | null;
  energy_min_eV?: number | null;
  energy_max_eV?: number | null;
  num_energy_points?: number | null;
};

function parsePointsCount(text: string, axisName: string): number | null {
  const pattern = new RegExp(
    `${axisName}\\s*=\\s*\\{[^}]*Points\\s*=\\s*\\(\\s*(\\d+)`,
    "s",
  );
  const match = pattern.exec(text);
  return match ? Number.parseInt(match[1] ?? "", 10) : null;
}

function parsePointsArray(text: string, axisName: string): number[] | null {
  const pattern = new RegExp(
    `${axisName}\\s*=\\s*\\{[^}]*Points\\s*=\\s*\\(\\s*\\d+\\s*,\\s*([\\d\\s.,\\-eE+]+)\\)`,
    "s",
  );
  const match = pattern.exec(text);
  if (!match?.[1]) {
    return null;
  }
  return match[1]
    .replace(/,/g, " ")
    .split(/\s+/)
    .filter(Boolean)
    .map((value) => Number.parseFloat(value));
}

function parseHdrScalar(raw: string, field: string): number | null {
  const number = String.raw`[-+]?(?:\d+\.?\d*|\.\d+)(?:[eE][-+]?\d+)?`;
  const pattern = new RegExp(String.raw`^\s*${field}\s*=\s*(${number})\s*(?:$|\#)`, "im");
  const match = pattern.exec(raw);
  return match?.[1] ? Number.parseFloat(match[1]) : null;
}

function parseHdrInt(raw: string, field: string): number | null {
  const pattern = new RegExp(String.raw`^\s*${field}\s*=\s*(\d+)\s*(?:$|\#)`, "im");
  const match = pattern.exec(raw);
  return match?.[1] ? Number.parseInt(match[1], 10) : null;
}

function axisIsEnergy(name: string): boolean {
  const lowered = name.toLowerCase();
  return lowered.includes("energy") || lowered.endsWith("ev") || lowered.includes("photon");
}

function energyAxisValues(meta: HdrMeta): number[] | null {
  const pName = meta.paxis_name ?? "";
  const qName = meta.qaxis_name ?? "";
  const pPoints = meta.paxis_points;
  const qPoints = meta.qaxis_points;
  if (axisIsEnergy(pName) && pPoints) {
    return pPoints;
  }
  if (axisIsEnergy(qName) && qPoints) {
    return qPoints;
  }
  return null;
}

/**
 * Derive catalog energy metadata from a parsed STXM header dict.
 */
export function extractScanEnergy(meta: Pick<HdrMeta, "raw"> & Partial<HdrMeta>): {
  energy_eV: number | null;
  energy_min_eV: number | null;
  energy_max_eV: number | null;
  num_energy_points: number | null;
} {
  const empty = {
    energy_eV: null,
    energy_min_eV: null,
    energy_max_eV: null,
    num_energy_points: null,
  };
  const axisValues = energyAxisValues(meta as HdrMeta);
  if (axisValues && axisValues.length > 0) {
    const lo = Math.min(...axisValues);
    const hi = Math.max(...axisValues);
    const count = axisValues.length;
    if (count === 1 || Math.abs(lo - hi) < 1e-6) {
      return {
        energy_eV: lo,
        energy_min_eV: lo,
        energy_max_eV: hi,
        num_energy_points: count,
      };
    }
    return {
      energy_eV: null,
      energy_min_eV: lo,
      energy_max_eV: hi,
      num_energy_points: count,
    };
  }
  const raw = meta.raw;
  let energy = parseHdrScalar(raw, "Energy");
  energy ??= parseHdrScalar(raw, "PhotonEnergy");
  const start = parseHdrScalar(raw, "StartEnergy");
  const end = parseHdrScalar(raw, "EndEnergy");
  const numEnergy = parseHdrInt(raw, "NumEnergy");
  if (energy !== null) {
    return {
      energy_eV: energy,
      energy_min_eV: energy,
      energy_max_eV: energy,
      num_energy_points: numEnergy ?? 1,
    };
  }
  if (start !== null && end !== null) {
    const lo = Math.min(start, end);
    const hi = Math.max(start, end);
    const count = numEnergy;
    if (Math.abs(lo - hi) < 1e-6) {
      return {
        energy_eV: lo,
        energy_min_eV: lo,
        energy_max_eV: hi,
        num_energy_points: count ?? 1,
      };
    }
    return {
      energy_eV: null,
      energy_min_eV: lo,
      energy_max_eV: hi,
      num_energy_points: count,
    };
  }
  if (start !== null) {
    return {
      energy_eV: start,
      energy_min_eV: start,
      energy_max_eV: start,
      num_energy_points: numEnergy ?? 1,
    };
  }
  if (numEnergy !== null && numEnergy > 1) {
    return {
      energy_eV: null,
      energy_min_eV: null,
      energy_max_eV: null,
      num_energy_points: numEnergy,
    };
  }
  return empty;
}

/**
 * Parse STXM .hdr file metadata including axis sizes and optional point arrays.
 */
export function readHdr(hdrPath: string): HdrMeta {
  const raw = fs.readFileSync(hdrPath, "utf8");
  const paxisCount = parsePointsCount(raw, "PAxis");
  const qaxisCount = parsePointsCount(raw, "QAxis");
  if (paxisCount === null || qaxisCount === null) {
    throw new Error("Could not find PAxis or QAxis Points in header");
  }
  const out: HdrMeta = {
    paxis_count: paxisCount,
    qaxis_count: qaxisCount,
    raw,
  };
  const pname = /PAxis\s*=\s*\{\s*Name\s*=\s*"([^"]*)"/.exec(raw);
  const qname = /QAxis\s*=\s*\{\s*Name\s*=\s*"([^"]*)"/.exec(raw);
  if (pname?.[1]) {
    out.paxis_name = pname[1];
  }
  if (qname?.[1]) {
    out.qaxis_name = qname[1];
  }
  const parr = parsePointsArray(raw, "PAxis");
  const qarr = parsePointsArray(raw, "QAxis");
  if (parr) {
    out.paxis_points = parr;
  }
  if (qarr) {
    out.qaxis_points = qarr;
  }
  Object.assign(out, extractScanEnergy(out));
  return out;
}

/**
 * Load STXM .xim ascii image as a 2D float array (rows x cols).
 */
export function readXim(ximPath: string, shape?: [number, number]): number[][] {
  const text = fs.readFileSync(ximPath, "utf8").trim();
  const rows = text.split(/\r?\n/).filter((line) => line.trim().length > 0);
  const data = rows.map((line) =>
    line
      .trim()
      .split(/\s+/)
      .map((value) => Number.parseFloat(value)),
  );
  if (data.length === 1 && shape) {
    const [nRows, nCols] = shape;
    const flat = data[0] ?? [];
    if (flat.length !== nRows * nCols) {
      throw new Error(`xim flat length ${flat.length} != ${nRows * nCols}`);
    }
    const reshaped: number[][] = [];
    for (let row = 0; row < nRows; row += 1) {
      reshaped.push(flat.slice(row * nCols, (row + 1) * nCols));
    }
    return reshaped;
  }
  if (shape) {
    const [nRows, nCols] = shape;
    const total = data.reduce((sum, row) => sum + row.length, 0);
    if (total === nRows * nCols && (data.length !== nRows || (data[0]?.length ?? 0) !== nCols)) {
      const flat = data.flat();
      const reshaped: number[][] = [];
      for (let row = 0; row < nRows; row += 1) {
        reshaped.push(flat.slice(row * nCols, (row + 1) * nCols));
      }
      return reshaped;
    }
  }
  return data;
}

function resolveXimPath(hdrPath: string): string {
  const parent = path.dirname(hdrPath);
  const stem = path.basename(hdrPath, path.extname(hdrPath));
  let ximPath = path.join(parent, `${stem}_a.xim`);
  if (!fs.existsSync(ximPath)) {
    ximPath = path.join(parent, `${stem}.xim`);
  }
  if (!fs.existsSync(ximPath)) {
    throw new Error(`xim file not found: ${ximPath}`);
  }
  return ximPath;
}

/**
 * Load STXM scan: parse .hdr and load associated .xim ascii image.
 */
export function loadStxm(
  hdrPath: string,
  ximPath?: string,
): { meta: HdrMeta; image: number[][] } {
  const meta = readHdr(hdrPath);
  const resolvedXim = ximPath ?? resolveXimPath(hdrPath);
  const shape: [number, number] = [meta.qaxis_count, meta.paxis_count];
  const image = readXim(resolvedXim, shape);
  return { meta, image };
}

/**
 * Return true if the .hdr header contains Type = "NEXAFS Line Scan".
 */
export function isNexafsLineScanType(hdrPath: string): boolean {
  if (!hdrPath.toLowerCase().endsWith(".hdr") || !fs.existsSync(hdrPath)) {
    return false;
  }
  return fs.readFileSync(hdrPath, "utf8").includes(NEXAFS_LINE_SCAN_TYPE);
}

/**
 * Fast heuristic: header parses and .xim has sufficient byte size.
 */
export function isValidLineScanFast(hdrPath: string): boolean {
  try {
    const meta = readHdr(hdrPath);
    const n = meta.paxis_count * meta.qaxis_count;
    const ximPath = resolveXimPath(hdrPath);
    return fs.statSync(ximPath).size >= n * MIN_BYTES_PER_VALUE;
  } catch {
    return false;
  }
}

/**
 * Return true if loadStxm succeeds for the given header.
 */
export function isValidLineScan(hdrPath: string): boolean {
  try {
    loadStxm(hdrPath);
    return true;
  } catch {
    return false;
  }
}

/**
 * Return true if header is NEXAFS Line Scan and loads with correct 2D shape.
 */
export function isNexafsLineScan(hdrPath: string): boolean {
  return isNexafsLineScanType(hdrPath) && isValidLineScan(hdrPath);
}

/**
 * Read the STXM Type field from a .hdr file.
 */
export function parseHdrScanType(hdrPath: string): string {
  if (!fs.existsSync(hdrPath)) {
    return "Unknown";
  }
  const match = HDR_TYPE_PATTERN.exec(fs.readFileSync(hdrPath, "utf8"));
  return match?.[1] ?? "Unknown";
}

/**
 * Map a header Type string to a stable UI grouping key.
 */
export function scanTypeCategory(scanType: string): ScanCategory {
  const lowered = scanType.toLowerCase();
  if (lowered.includes("nexafs line scan") || lowered.includes("line scan")) {
    return "line_scan";
  }
  if (lowered.includes("image scan")) {
    return "image_scan";
  }
  if (lowered.includes("focus scan")) {
    return "focus_scan";
  }
  if (lowered.includes("fixed point") || lowered.includes("fixed-point")) {
    return "fixed_point";
  }
  if (lowered.includes("stack")) {
    return "stack";
  }
  return "other";
}

/**
 * List .hdr files under an experiment directory tree, sorted by path.
 */
export function listExperimentHdrFiles(experimentPath: string): string[] {
  if (!fs.existsSync(experimentPath) || !fs.statSync(experimentPath).isDirectory()) {
    return [];
  }
  const paths: string[] = [];
  const walk = (dir: string): void => {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(full);
      } else if (entry.isFile() && entry.name.toLowerCase().endsWith(".hdr")) {
        try {
          const resolved = fs.realpathSync.native(full);
          if (fs.existsSync(resolved)) {
            paths.push(resolved);
          }
        } catch {
          continue;
        }
      }
    }
  };
  walk(experimentPath);
  return paths.sort();
}

/**
 * List NEXAFS line scans that pass isNexafsLineScan, sorted by name.
 */
export function listNexafsLineScans(experimentPath: string): string[] {
  if (!fs.existsSync(experimentPath) || !fs.statSync(experimentPath).isDirectory()) {
    return [];
  }
  return listExperimentHdrFiles(experimentPath)
    .filter((hdrPath) => isNexafsLineScan(hdrPath))
    .sort((a, b) => path.basename(a).localeCompare(path.basename(b)));
}
